File: heartwood-mcp/tools/locations.py
"""Location tools — create and query JobTread locations."""

from __future__ import annotations

from app import mcp
from constants import ORG_ID
from pave import call_pave, flatten_nodes


@mcp.tool()
async def jt_create_location(
    account_id: str,
    street1: str,
    city: str,
    state: str,
    zip_code: str,
    street2: str | None = None,
) -> dict:
    """Create a location (address) linked to an account.

    Args:
        account_id: The parent account ID.
        street1: Street address line 1.
        city: City.
        state: State abbreviation.
        zip_code: ZIP/postal code.
        street2: Street address line 2 (optional).
    """
    address: dict = {
        "street1": street1,
        "city": city,
        "state": state,
        "zip": zip_code,
    }
    if street2:
        address["street2"] = street2

    return await call_pave(
        "createLocation",
        params={
            "organizationId": ORG_ID,
            "accountId": account_id,
            "address": address,
        },
        fields={
            "id": {},
            "address": {"street1": {}, "street2": {}, "city": {}, "state": {}, "zip": {}},
        },
    )


@mcp.tool()
async def jt_get_locations(
    account_id: str | None = None,
    first: int = 100,
) -> list[dict]:
    """List locations, optionally filtered by account.

    Args:
        account_id: Filter by parent account.
        first: Max results.
    """
    filter_obj: dict = {}
    if account_id:
        filter_obj["accountId"] = {"eq": account_id}

    params: dict = {"first": first}
    if filter_obj:
        params["filter"] = filter_obj

    result = await call_pave(
        "locations",
        params=params,
        fields={
            "nodes": {
                "id": {},
                "accountId": {},
                "address": {"street1": {}, "street2": {}, "city": {}, "state": {}, "zip": {}},
            }
        },
    )
    return flatten_nodes(result)

File: workspace/projects/heartwood-mcp/src/tools/jt/locations.ts
/**
 * JT Location tools — 2 tools
 */

import type { PaveClient, ToolResult } from "../../pave/index.js";
import { LOCATION_FIELDS } from "../../pave/index.js";
import type { ToolDef } from "../registry.js";

export function locationTools(pave: PaveClient): ToolDef[] {
  return [
    // ── jt_create_location ─────────────────────────────────────────────
    {
      name: "jt_create_location",
      description:
        "Create a location for an account. Required before creating a job.",
      inputSchema: {
        type: "object" as const,
        properties: {
          accountId: { type: "string", description: "Parent account ID" },
          address: { type: "string", description: "Street address" },
          name: { type: "string", description: "Location name" },
          city: { type: "string", description: "City (optional)" },
          state: { type: "string", description: "State (optional)" },
          zip: { type: "string", description: "ZIP code (optional)" },
          contactId: { type: "string", description: "Associated contact ID (optional)" },
        },
        required: ["accountId", "address", "name"],
      },
      handler: async (params: Record<string, unknown>): Promise<ToolResult> => {
        const data: Record<string, unknown> = {
          accountId: params.accountId,
          address: params.address,
          name: params.name,
        };
        if (params.city) data.city = params.city;
        if (params.state) data.state = params.state;
        if (params.zip) data.zip = params.zip;
        if (params.contactId) data.contactId = params.contactId;
        return pave.create("location", data, LOCATION_FIELDS);
      },
    },

    // ── jt_get_locations ───────────────────────────────────────────────
    {
      name: "jt_get_locations",
      description: "Get locations, optionally filtered by account.",
      inputSchema: {
        type: "object" as const,
        properties: {
          accountId: { type: "string", description: "Filter by account ID (optional)" },
        },
        required: [],
      },
      handler: async (params: Record<string, unknown>): Promise<ToolResult> => {
        const filter = params.accountId
          ? { conditions: [{ field: "accountId", operator: "eq" as const, value: params.accountId }] }
          : undefined;
        return pave.query({
          entity: "location",
          fields: LOCATION_FIELDS,
          filter,
        });
      },
    },
  ];
}
